tokio/src/sync/mpsc/bounded.rs

```diff
@@ -849,10 +861,68 @@ impl<T> Sender<T> {
     /// }
     /// ```
     pub async fn reserve(&self) -> Result<Permit<'_, T>, SendError<()>> {
-        self.reserve_inner().await?;
+        self.reserve_inner(1).await?;
         Ok(Permit { chan: &self.chan })
     }
 
+    /// Waits for channel capacity. Once capacity to send `n` messages is
+    /// available, it is reserved for the caller.
+    ///
+    /// If the channel is full, the function waits for the number of unreceived
+    /// messages to become `n` less than the channel capacity. Capacity to send `n`
+    /// message is reserved for the caller. A [`PermitIterator`] is returned to track
+    /// the reserved capacity. You can call this [`Iterator`] until it is exhausted to
+    /// get a [`Permit`] and then call [`Permit::send`].
+    ///
+    /// Dropping [`PermitIterator`] without sending all messages releases the capacity back
+    /// to the channel.
+    ///
+    /// [`PermitIterator`]: PermitIterator
+    /// [`Permit`]: Permit
+    /// [`send`]: Permit::send
+    ///
+    /// # Cancel safety
+    ///
+    /// This channel uses a queue to ensure that calls to `send` and `reserve`
+    /// complete in the order they were requested. Cancelling a call to
+    /// `reserve` makes you lose your place in the queue.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::channel(2);
+    ///
+    ///     // Reserve capacity
+    ///     let mut permit = tx.reserve_many(2).await.unwrap();
+    ///
+    ///     // Trying to send directly on the `tx` will fail due to no
+    ///     // available capacity.
+    ///     assert!(tx.try_send(123).is_err());
+    ///
+    ///     // Sending with the permit iterator succeeds
+    ///     permit.next().unwrap().send(456);
+    ///     permit.next().unwrap().send(457);
+    ///
+    ///     // The iterator should now be exhausted
+    ///     assert!(permit.next().is_none());
+    ///     
+    ///     // The value sent on the permit is received
+    ///     assert_eq!(rx.recv().await.unwrap(), 456);
+    ///     assert_eq!(rx.recv().await.unwrap(), 457);
+    /// }
+    /// ```
+    pub async fn reserve_many(&self, n: usize) -> Result<PermitIterator<'_, T>, SendError<()>> {
+        self.reserve_inner(n).await?;
+        Ok(PermitIterator {
+            chan: &self.chan,
+            n,
+        })
+    }
```

Comment:
I would return a `SendError` if `n > batch_semaphore::MAX_PERMITS` here. I'm not sure how the batch semaphore handles it if you pass something greater than that.

Reply:
Indeed, the batch semaphore panics if n > MAX_PERMITS. I added a check to fix this.

Reply:
Maybe we should also return an error if `n` is greater than the channel's max capacity? If we allow it, the channel will be blocked forever for all senders due to the semaphore's fairness.

Reply:
Then the MAX_PERMIT check is not required because the capacity is always < MAX_PERMIT

Reply:
For `reserve_many`, I think returning an error makes more sense than a panic, since if you exceed MAX_PERMITS, then you are necessarily also exceeding the number of available permits.